tokens/__init__.py

```diff
@@ -96,19 +107,30 @@ def refresh(token_name):
 
     r = requests.post(url, data=body, auth=auth)
     r.raise_for_status()
-    data = r.json()
-    token['data'] = data
-    token['expires_at'] = time.time() + data.get('expires_in')
-    token['access_token'] = data.get('access_token')
+    try:
+        data = r.json()
+        token['data'] = data
+        token['expires_at'] = time.time() + data['expires_in']
+        token['access_token'] = data['access_token']
+    except Exception as e:
+        raise InvalidTokenResponse('Expected a JSON object with keys "expires_in" and "access_token": {}'.format(e))
+    if not token['access_token']:
+        raise InvalidTokenResponse('Empty "access_token" value')
     return token
 
 
 def get(token_name):
     token = TOKENS[token_name]
     access_token = token.get('access_token')
-    # TODO: remove hardcoded time value (20 minutes)
-    if not access_token or time.time() > token['expires_at'] - 20*60:
-        token = refresh(token_name)
+    if not access_token or time.time() > token['expires_at'] - REFRESH_BEFORE_SECS_LEFT:
+        try:
+            token = refresh(token_name)
+        except Exception as e:
+            if access_token and time.time() < token['expires_at'] + EXPIRATION_TOLERANCE_SECS:
+                # apply some tolerance, still try our old token if it's still valid
+                logger.warn('Failed to refresh access token "%s" (but it is still valid): %s', token_name, e)
+            else:
+                raise
 
     access_token = token.get('access_token')
```

Comment:
This assignment is not needed in the Exception case as it will be equivalent to the assignment on line 124, so maybe it can moved to line 128 to make it clear it only changes if the refresh is successful.